File: src/utils/drag_particles.cpp
#include "drag_particles.hpp"
#include <GLFW/glfw3.h>
#include <glm/glm.hpp>
#include "../classes/particle.hpp"
#include <memory>
#include "../config.hpp"
#include "../classes/camera.hpp"
#include "../classes/simulation.hpp"
#include "ray.hpp"

DragParticles::DragParticles(GLFWwindow* window, bool fixedDrag) {
    this->window = window;
    this->fixedDrag = fixedDrag;
}

void DragParticles::setDraggedParticle(std::shared_ptr<Sphere> particle) {
    draggedParticle = particle;
    particle->setUpdatingEnabled(false);
}

void DragParticles::unsetDraggedParticle() {
    if (draggedParticle) {
        draggedParticle->setUpdatingEnabled(true);
    }
    draggedParticle = nullptr;
}

void DragParticles::setDragDistance(float dragDistance) {
    this->dragDistance = dragDistance;
}

void DragParticles::setWindow(GLFWwindow* window) {
    this->window = window;
}

void DragParticles::setLastMousePos(glm::vec2 pos) {
    lastMousePos = pos;
}

void DragParticles::setIsDragging(bool isDragging) {
    this->isDragging = isDragging;
}

void DragParticles::handleDrag(const Camera &camera, Simulation &simulation) {
    if (glfwGetMouseButton(window, GLFW_MOUSE_BUTTON_LEFT) == GLFW_PRESS) {
        double x, y;
        glfwGetCursorPos(window, &x, &y);
        glm::vec2 mousePos = glm::vec2(x, y);

        // Get the camera position and view matrix  
        glm::vec3 cameraPosition = camera.position;
        glm::mat4 viewMatrix = camera.getViewMatrix();
        glm::mat4 projectionMatrix = camera.getProjectionMatrix(45.0f, (float)WINDOW_WIDTH / (float)WINDOW_HEIGHT, 0.1f, 100.0f);

        if (!isDragging) {
            isDragging = true;
            lastMousePos = mousePos;

            // handle getting the particle to drag


            // Convert the 2D mouse position to normalized device coordinates
            glm::vec4 ray_clip = glm::vec4((2.0f * mousePos.x) / WINDOW_WIDTH - 1.0f, 1.0f - (2.0f * mousePos.y) / WINDOW_HEIGHT, -1.0f, 1.0f);
            glm::vec4 ray_eye = glm::inverse(projectionMatrix) * ray_clip;
            ray_eye = glm::vec4(ray_eye.x, ray_eye.y, -1.0f, 0.0f);
            glm::vec3 ray_world = glm::vec3(glm::inverse(viewMatrix) * ray_eye);

            // Create a ray from the camera position in the direction of the mouse click
            Ray ray(cameraPosition, glm::normalize(ray_world));

            // Check for intersections between the ray and the particles
            std::shared_ptr<Sphere> selectedParticle = nullptr;
            float minDistance = std::numeric_limits<float>::max();
            for (auto& particle : simulation.spheres) {
                float distance = ray.intersect(particle);
                if (distance < minDistance) {
                    selectedParticle = particle;
                    minDistance = distance;
                }
            }

            // If there's an intersection, set the intersected particle as the dragged particle
            if (selectedParticle) {
                setDraggedParticle(selectedParticle);
                setDragDistance(minDistance);
            }

        }

        if (draggedParticle) {
            if (!fixedDrag) {
                // Convert the 2D mouse position difference to a 3D movement
                glm::vec2 mousePosDiff = mousePos - lastMousePos;
                glm::vec3 movement = glm::vec3(mousePosDiff.x, -mousePosDiff.y, 0.0f); // Invert y-axis because screen coordinates are inverted

                // Calculate the distance from the camera to the particle
                float distance = glm::length(cameraPosition - draggedParticle->position);

                // Calculate a scale factor based on the distance
                float scaleFactor = distance / 10000.0f; // 10000.0f is an arbitrary value that works well for this purpose

                // Scale the movement to match your scene dimensions and the distance from the camera
                movement *= scaleFactor;

                // Apply the movement to the dragged particle
                draggedParticle->move(movement);

                // Update the last mouse position
                lastMousePos = mousePos;
            } else {
                // Convert the 2D mouse position to normalized device coordinates
                glm::vec4 ray_clip = glm::vec4((2.0f * mousePos.x) / WINDOW_WIDTH - 1.0f, 1.0f - (2.0f * mousePos.y) / WINDOW_HEIGHT, 1.0f, 1.0f);
                glm::vec4 ray_eye = glm::inverse(projectionMatrix) * ray_clip;
                ray_eye = glm::vec4(ray_eye.x, ray_eye.y, -1.0f, 0.0f);
                glm::vec3 ray_world = glm::normalize(glm::vec3(glm::inverse(viewMatrix) * ray_eye));

                // Calculate the new position of the particle
                glm::vec3 newPosition = camera.position + ray_world * dragDistance;

                // Move the particle to the new position
                draggedParticle->position = newPosition;

                // Update the last mouse position
                lastMousePos = mousePos;
            }
        }


    } else { // on release
        isDragging = false;
        unsetDraggedParticle();
    }
}

File: rust/src/utils/drag_particles.rs
use crate::classes::camera::Camera;
use crate::classes::simulation::Simulation;
use crate::classes::{particle::ParticleTrait, particles::sphere::Sphere};
use crate::utils::ray::Ray;
use glium::winit::event::{ElementState, Event, KeyEvent, MouseButton, WindowEvent};
use glium::winit::keyboard::{KeyCode, PhysicalKey};
use nalgebra::{Vector3, Vector4};
use std::f32::INFINITY;
use std::sync::{Arc, RwLock};

use crate::config::{
    CAMERA_ASPECT_RATIO, CAMERA_FAR, CAMERA_FOV, CAMERA_NEAR, WINDOW_HEIGHT, WINDOW_WIDTH,
};

pub struct DragState {
    pub left_mouse_button_pressed: bool,
    pub last_cursor_pos: Option<(f64, f64)>,
    pub current_cursor_pos: Option<(f64, f64)>,
}

impl DragState {
    pub fn new() -> Self {
        DragState {
            left_mouse_button_pressed: false,
            last_cursor_pos: None,
            current_cursor_pos: None,
        }
    }
}

pub struct DragParticles {
    last_cursor_pos: (f32, f32),
    is_dragging: bool,
    dragged_particle: Option<Arc<RwLock<Sphere>>>,
    drag_distance: f32,
    fixed_drag: bool,
}

impl DragParticles {
    pub fn new(fixed_drag: bool) -> Self {
        DragParticles {
            last_cursor_pos: (0., 0.),
            is_dragging: false,
            dragged_particle: None,
            drag_distance: 0.0,
            fixed_drag: fixed_drag,
        }
    }

    pub fn set_dragged_particle(&mut self, particle: Arc<RwLock<Sphere>>) {
        particle.write().unwrap().set_updating_enabled(false);
        self.dragged_particle = Some(particle);
    }

    pub fn unset_dragged_particle(&mut self) {
        if let Some(particle) = &self.dragged_particle {
            particle.write().unwrap().set_updating_enabled(true);
        }
        self.dragged_particle = None;
    }

    pub fn set_drag_distance(&mut self, distance: f32) {
        self.drag_distance = distance;
    }

    pub fn set_is_dragging(&mut self, is_dragging: bool) {
        self.is_dragging = is_dragging;
    }

    pub fn process_drag(&mut self, camera: &Arc<RwLock<Camera>>, sim: &mut Simulation, drag_state: &Arc<RwLock<DragState>>) {
        if drag_state.read().unwrap().left_mouse_button_pressed {
            // supposing we are currently pressing the left mouse button

            let camera_read = camera.read().unwrap();
            let drag_state_read = drag_state.read().unwrap();

            let camera_position = camera_read.position;
            let view_matrix = camera_read.get_view_matrix();
            let projection_matrix = camera_read.get_projection_matrix(CAMERA_FOV, CAMERA_ASPECT_RATIO, CAMERA_NEAR, CAMERA_FAR);

            if !self.is_dragging {
                self.is_dragging = true;
                self.last_cursor_pos = (
                    drag_state_read.current_cursor_pos.unwrap().0 as f32,
                    drag_state_read.current_cursor_pos.unwrap().1 as f32,
                );

                let ray_clip = Vector4::new(
                    (2.0 * self.last_cursor_pos.0 / WINDOW_WIDTH) - 1.0,
                    1.0 - (2.0 * self.last_cursor_pos.1 / WINDOW_HEIGHT),
                    -1.0,
                    1.0,
                );
                let mut ray_eye: Vector4<f32>;
                if let Some(inverse_projection_matrix) = projection_matrix.try_inverse() {
                    ray_eye = inverse_projection_matrix * ray_clip;
                    ray_eye = Vector4::new(ray_eye.x, ray_eye.y, -1.0, 0.0);
                } else {
                    eprintln!("Failed to invert the projection matrix.");
                    return;
                }

                let ray_world: Vector3<f32>;
                if let Some(inverse_view_matrix) = view_matrix.try_inverse() {
                    let ray_world_in = inverse_view_matrix * ray_eye;
                    ray_world = Vector3::new(ray_world_in.x, ray_world_in.y, ray_world_in.z);
                } else {
                    eprintln!("Failed to invert the view matrix.");
                    return;
                }

                let ray = Ray::new(camera_position, ray_world.normalize());

                let mut selected_particle = None;
                let mut min_distance = INFINITY;
                for sphere in sim.spheres.iter() {
                    let distance = ray.intersect(sphere.clone());
                    if distance < min_distance {
                        selected_particle = Some(sphere.clone());
                        min_distance = distance;
                    }
                }

                if let Some(particle) = selected_particle {
                    self.set_dragged_particle(particle);
                    self.set_drag_distance(min_distance);
                }
            }

            if let Some(dragged_particle) = self.dragged_particle.as_ref() {
                if !self.fixed_drag {
                    let mouse_pos_diff = (
                        drag_state_read.current_cursor_pos.unwrap().0 as f32 - self.last_cursor_pos.0,
                        drag_state_read.current_cursor_pos.unwrap().1 as f32 - self.last_cursor_pos.1,
                    );

                    let mut dragged_particle = dragged_particle.write().unwrap();

                    let mut movement = Vector3::new(mouse_pos_diff.0, -mouse_pos_diff.1, 0.0);

                    let distance = (camera_position - dragged_particle.get_position()).norm();

                    let scale_factor = distance / 10000.0;

                    movement *= scale_factor;

                    dragged_particle.move_by(movement);

                    self.last_cursor_pos = (
                        drag_state_read.current_cursor_pos.unwrap().0 as f32,
                        drag_state_read.current_cursor_pos.unwrap().1 as f32,
                    );
                } else {
                    let ray_clip = Vector4::new(
                        (2.0 * self.last_cursor_pos.0 / WINDOW_WIDTH) - 1.0,
                        1.0 - (2.0 * self.last_cursor_pos.1 / WINDOW_HEIGHT),
                        -1.0,
                        1.0,
                    );
                    let mut ray_eye: Vector4<f32>;
                    if let Some(inverse_projection_matrix) = projection_matrix.try_inverse() {
                        ray_eye = inverse_projection_matrix * ray_clip;
                        ray_eye = Vector4::new(ray_eye.x, ray_eye.y, -1.0, 0.0);
                    } else {
                        eprintln!("Failed to invert the projection matrix.");
                        return;
                    }
    
                    let ray_world: Vector3<f32>;
                    if let Some(inverse_view_matrix) = view_matrix.try_inverse() {
                        let ray_world_in = inverse_view_matrix * ray_eye;
                        ray_world = Vector3::new(ray_world_in.x, ray_world_in.y, ray_world_in.z).normalize();
                    } else {
                        eprintln!("Failed to invert the view matrix.");
                        return;
                    }

                    let new_position = camera_position + ray_world * self.drag_distance;

                    dragged_particle.write().unwrap().set_position(new_position);

                    self.last_cursor_pos = (
                        drag_state_read.current_cursor_pos.unwrap().0 as f32,
                        drag_state_read.current_cursor_pos.unwrap().1 as f32,
                    );
                }
            }
        } else {
            self.is_dragging = false;
            self.unset_dragged_particle();
        }
    }

    pub fn handle_drag_controls(&self, event: &WindowEvent, last_cursor_pos: &mut Option<(f64, f64)>, drag_state: &Arc<RwLock<DragState>>) {
        match event {
            WindowEvent::CursorMoved { position, .. } => {
                let (x, y) = (position.x, position.y);
                if let Some((last_x, last_y)) = last_cursor_pos {
                    let dx = x - *last_x;
                    let dy = y - *last_y;

                    let mut drag_state_write = drag_state.write().unwrap();
                    drag_state_write.last_cursor_pos = drag_state_write.current_cursor_pos;
                    drag_state_write.current_cursor_pos = Some((x, y));

                    // process drag in main loop
                }
                *last_cursor_pos = Some((x, y));
            }
            WindowEvent::MouseInput { state, button, .. } => {
                if *button == MouseButton::Left {
                    let mut drag_state = drag_state.write().unwrap();
                    drag_state.left_mouse_button_pressed = match state {
                        ElementState::Pressed => true,
                        ElementState::Released => false,
                    };
                }
            }
            _ => (),
        }
    }
}
